### PR TITLE
SCTP: Fix type specs for inet:getopts/2 and inet:setopts/2

### PR DESCRIPTION
**Pull request by dotsimon**
SCTP Socket options have some weird[^1] semantics that this PR fixes.

Specifically:

- two options are read-only (see `prim_inet:type_opt/2`)
- many options for getopts require a  `#sctp_...` record to specify the association

Previously, the same type specification was used for both getopts and setopts which caused dialyzer to barf on perfectly valid code like this:
```erlang
    inet:getopts(Sock, [{sctp_status, #sctp_status{assoc_id = AssocId}}])
```

[^1]: Weird, as in not like TCP or UDP but perfectly fine for SCTP!

### CODE REVIEW

**Comment by dotsimon** [comment]
On master, the following additional changes would need to be made to accommodate the new connectx API:
```udiff
***************
*** 396,402 ****
        SockAddrs:: [{inet:ip_address(), inet:port_number()} |
                     inet:family_address() |
                     socket:sockaddr_in() | socket:sockaddr_in6()],
-       Opts     :: [option()].
  %%
  connectx_init(S, SockAddrs, Opts) ->
      case do_connectx(S, SockAddrs, Opts) of
--- 405,411 ----
        SockAddrs:: [{inet:ip_address(), inet:port_number()} |
                     inet:family_address() |
                     socket:sockaddr_in() | socket:sockaddr_in6()],
+       Opts     :: [setoption()].
  %%
  connectx_init(S, SockAddrs, Opts) ->
      case do_connectx(S, SockAddrs, Opts) of
***************
*** 411,417 ****
        Socket :: sctp_socket(),
        Addrs :: [inet:ip_address() | inet:hostname()],
        Port :: inet:port_number() | atom(),
-       Opts :: [option()].
  %%
  connectx_init(S, Addrs, Port, Opts) ->
      connectx_init(S, Addrs, Port, Opts, infinity).
--- 420,426 ----
        Socket :: sctp_socket(),
        Addrs :: [inet:ip_address() | inet:hostname()],
        Port :: inet:port_number() | atom(),
+       Opts :: [setoption()].
  %%
  connectx_init(S, Addrs, Port, Opts) ->
      connectx_init(S, Addrs, Port, Opts, infinity).
***************
*** 421,427 ****
        Socket :: sctp_socket(),
        Addrs :: [inet:ip_address() | inet:hostname()],
        Port :: inet:port_number() | atom(),
-       Opts :: [option()],
        Timeout :: timeout().
  %%
  connectx_init(S, Addrs, Port, Opts, Timeout) ->
--- 430,436 ----
        Socket :: sctp_socket(),
        Addrs :: [inet:ip_address() | inet:hostname()],
        Port :: inet:port_number() | atom(),
+       Opts :: [setoption()],
        Timeout :: timeout().
  %%
  connectx_init(S, Addrs, Port, Opts, Timeout) ->

```

**Comment by dotsimon** [comment]
In the documentation I removed the references to the option type because it was only partially expanded
```
setoption() = elementary_option() | record_option()
```
Which did not seem useful, especially since all the options are explicitly defined in the _SCTP Socket Options_ section

**Comment by dotsimon** [comment]
Regarding the automated check:
I don't understand why the "Build Erlang/OTP (documentation)" action fails.
The doc builds ok locally and also in the "Build Erlang/OTP (32-bit) (pull_request)" action.

**Comment by garazdawi** [comment]
It is the xml lint that fails, if you run `make xmllint` on the top you should be able to reproduce it locally. The reason why it fails is because you have removed the option type, but somewhere in the docs it is still being used.

**Comment by garazdawi** [comment]
https://github.com/erlang/otp/blob/master/lib/kernel/doc/src/gen_sctp.xml#L529 this seems to be the place where it is used. It should probably be replaced by `<seeerl marker="#options"`.

File: lib/kernel/doc/src/gen_sctp.xml
Comment (garazdawi) [inline lib/kernel/doc/src/gen_sctp.xml:87]: You need to add `getoption` in the list of documented types as the inet docs now refer to it.

File: lib/kernel/src/inet.erl
Comment (garazdawi) [inline lib/kernel/src/inet.erl:149]: This type needs to be inlined into the `getopts/2` documentation or documented at the top like `socket_setopt`.

**Review by garazdawi — DISMISSED** [review]

**Comment by dotsimon** [comment]
Is there some way to recreate what the check is doing?
It's kind of frustrating since a local build of the documentation succeeds albeit with some dangling links.

File: lib/kernel/doc/src/gen_sctp.xml
Comment (dotsimon) [reply lib/kernel/doc/src/gen_sctp.xml:87]: Ok yes I see that now, and also the optionval (return from inet:getopts) case needs to be fixed.
Although, I'm not sure what the best way to handle this is.
The type specs in gen_sctp now match what the code does.
But they are not so great for automatically generating the doc.
As I mentioned earlier, just to include `setoption` in the xml results in an expansion like this:
```
setoption() = elementary_option() | record_option()
```
Of course I could include all those types but it feels like that would just bloat the doc without really adding anything. Especially since the same would also apply for `getoption()` and `optionval()`

I'd be interested in what approach you suggest. I assume there is not a way to get the docgen to expand certain types?

**Comment by garazdawi** [comment]
I created #5946 to expand on how to do the documentation validation locally.

File: lib/kernel/doc/src/gen_sctp.xml
Comment (garazdawi) [reply lib/kernel/doc/src/gen_sctp.xml:87]: This is one of the bad parts of having the typespecs as the source for the documentation. Sometimes you need to make compromises either in the type annotations, or the documentation. I'll leave the recommendations about what to do to the people responsible for inet/sctp as I don't really know my way around these parts.

> I assume there is not a way to get the docgen to expand certain types?

Not that I know of.

File: lib/kernel/src/inet.erl
Comment (dotsimon) [reply lib/kernel/src/inet.erl:149]: I expanded everything out.
It is now consistent with other documentation

**Comment by dotsimon** [comment]
> I created #5946 to expand on how to do the documentation validation locally.

I had already found the github action scripts to help myself. But this is great that it will now be documented for future reference.
Thanks for your time with this PR.

**Comment by dotsimon** [comment]
@garazdawi Hi! Do you need something else from me on this PR?

**Comment by garazdawi** [comment]
I don't know much about the inet/sctp API so you will have to wait for the people that do know for the final review. I only commented here to help out with how to write the correct xml for what you wanted to achieve.

**Comment by dotsimon** [comment]
Your change looks fine to me. I hadn't considered existing users of the gen_sctp types.

No worries about the delay, this is obviously a non-priority improvement.

**Review by RaimoNiskanen — APPROVED** [review]

**Pull request merged**
